packages/lodestar/src/network/gossip/gossipsub.ts

```diff
@@ -0,0 +1,140 @@
+        const objSubnet = this.deserializeGossipMessage(topic, message);
+        transformedObj = objSubnet.object;
+        isValid = isValid && await validatorFn(transformedObj, objSubnet.subnet);
+      } catch (err) {
+        isValid = false;
+        this.logger.error(`Cannot validate message from ${message.from}, topic ${message.topicIDs}, error: ${err}`);
+      }
+    }
+    if (isValid && transformedObj) {
+      this.transformedObjects.set(this.getKey(message), transformedObj);
+    }
+    return isValid;
+  }
+
+  public _emitMessage(topics: string[], message: IGossipMessage): void {
+    const subscribedTopics = super.getTopics();
+    topics.forEach((topic) => {
+      if (subscribedTopics.includes(topic)) {
+        const transformedObj = this.transformedObjects.get(this.getKey(message));
+        if (transformedObj) {
+          super.emit(topic, transformedObj);
+          this.transformedObjects.delete(this.getKey(message));
+        }
+      }
+    });
+  }
+
+
+  private getTopicValidator(topic: string): GossipMessageValidatorFn {
+    for(let subnet = 0; subnet < ATTESTATION_SUBNET_COUNT; subnet++) {
+      if (topic === getGossipTopic(GossipEvent.ATTESTATION_SUBNET, "ssz", new Map([["subnet", String(subnet)]]))) {
+        return this.validator.isValidIncomingCommitteeAttestation as GossipMessageValidatorFn;
+      }
+    }
+
+    let result: Function;
+    switch(topic) {
+      case getGossipTopic(GossipEvent.BLOCK, "ssz"):
+        result = this.validator.isValidIncomingBlock;
+        break;
+      case getGossipTopic(GossipEvent.ATTESTATION, "ssz"):
+        result =  this.validator.isValidIncomingUnaggregatedAttestation;
+        break;
+      case getGossipTopic(GossipEvent.AGGREGATE_AND_PROOF, "ssz"):
+        result =  this.validator.isValidIncomingAggregateAndProof;
+        break;
+      case getGossipTopic(GossipEvent.ATTESTER_SLASHING, "ssz"):
+        result =  this.validator.isValidIncomingAttesterSlashing;
+        break;
+      case getGossipTopic(GossipEvent.PROPOSER_SLASHING, "ssz"):
+        result = this.validator.isValidIncomingProposerSlashing;
+        break;
+      case getGossipTopic(GossipEvent.VOLUNTARY_EXIT, "ssz"):
+        result =  this.validator.isValidIncomingVoluntaryExit;
+        break;
+      default:
+        throw new Error(`No validator for topic ${topic}`); 
+    }
+    return result as GossipMessageValidatorFn;
+  }
+
+  private deserializeGossipMessage(topic: string, message: IGossipMessage): { object: GossipObject; subnet?: number} {
+    for(let subnet = 0; subnet < ATTESTATION_SUBNET_COUNT; subnet++) {
+      if (topic === getGossipTopic(GossipEvent.ATTESTATION_SUBNET, "ssz", new Map([["subnet", String(subnet)]]))) {
+        return {object: deserializeGossipMessage<Attestation>(message, this.config.types.Attestation), subnet};
+      }
+    }
+    let objType: AnySSZType;
+    switch(topic) {
+      case getGossipTopic(GossipEvent.BLOCK, "ssz"):
```

Comment:
Right now "ssz" is hardcoded everywhere, which is fine for now but will likely need to be changed. We should anticipate having gossip managers for different encoding schemed (like ssz_snappy needed for mainnet)

Reply:
I'm working on generalizing that in snappy compression PR